kubespawner/spawner.py

```diff
@@ -2409,44 +2385,61 @@ def _start_reflector(
         and a new one started (for recovering from possible errors).
         """
-        key = kind
-        ReflectorClass = reflector_class
 
-        def on_reflector_failure():
-            self.log.critical(
-                "%s reflector failed, halting Hub.",
-                key.title(),
-            )
-            sys.exit(1)
+        if self.enable_user_namespaces:
+            # Create one reflector for all namespaces.
+            # This requires binding ServiceAccount to ClusterRole.
+
+            def on_reflector_failure():
+                # If reflector cannot be started, halt the JH application.
+                self.log.critical(
+                    "%s reflector failed, halting Hub.",
+                    kind.title(),
+                )
+                sys.exit(1)
 
-        previous_reflector = self.__class__.reflectors.get(key)
+            async def catch_reflector_start(func):
+                try:
+                    await func
+                except Exception:
+                    self.log.exception(f"Reflector for {kind} failed to start.")
+                    sys.exit(1)
+
+        else:
+            # Create a dedicated reflector for each namespace.
+            # This allows JH to run pods in multiple namespaces without binding ServiceAccount to ClusterRole.
+
+            on_reflector_failure = None
+
+            async def catch_reflector_start(func):
+                # If reflector cannot be started (e.g. insufficient access rights, namespace cannot be found),
+                # just raise an exception instead halting the entire JH application.
+                try:
+                    await func
+                except Exception:
+                    self.log.exception(f"Reflector for {kind} failed to start.")
+                    raise
+
+        key = self._get_reflector_key(kind)
+        previous_reflector = self.__class__.reflectors.get(key, None)
 
         if replace or not previous_reflector:
-            self.__class__.reflectors[key] = ReflectorClass(
+            self.__class__.reflectors[key] = reflector_class(
                 parent=self,
                 namespace=self.namespace,
                 on_failure=on_reflector_failure,
                 **kwargs,
             )
-            f = asyncio.ensure_future(self.__class__.reflectors[key].start())
-
-            async def catch_reflector_start():
-                try:
-                    await f
-                except Exception:
-                    self.log.exception(f"Reflector for {kind} failed to start.")
-                    sys.exit(1)
-
-            asyncio.create_task(catch_reflector_start())
+            await catch_reflector_start(self.__class__.reflectors[key].start())
 
         if replace and previous_reflector:
             # we replaced the reflector, stop the old one
-            asyncio.ensure_future(previous_reflector.stop())
+            await previous_reflector.stop()
```

Comment:
I think maybe we don't need to await this, but only ensure its scheduled to be done, what do you think?

```suggestion
            asyncio.ensure_future(previous_reflector.stop())
```

Reply:
Fixed